MapleServer2/PacketHandlers/Game/MasteryHandler.cs

```diff
@@ -0,0 +1,168 @@
+                    }
+                    else
+                    {
+                        // send notice to player saying they haven't got enough materials
+                        session.SendNotice("You've run out of materials.");
+                    }
+
+                    break;
+                }
+                default:
+                    return;
+            }
+        }
+
+        private static void RemoveRequiredItemsFromInventory(GameSession session, RecipeMetadata recipe)
+        {
+            List<Item> playerInventoryItems = new(session.Player.Inventory.Items.Values);
+            List<RecipeItem> ingredients = recipe.GetIngredients();
+
+            for (int i = 0; i < ingredients.Count; i++)
+            {
+                RecipeItem ingredient = ingredients.ElementAt(i);
+                Item item = playerInventoryItems.FirstOrDefault(x => x.Id == ingredient.Id);
+                if (item == null)
+                {
+                    continue;
+                }
+
+                // check if whole stack will be used, and remove the item
+                // otherwise we want to just want to subtract the amount
+                if (ingredient.Amount == item.Amount)
+                {
+                    InventoryController.Remove(session, item.Uid, out Item _);
+                }
+                else
+                {
+                    InventoryController.Update(session, item.Uid, item.Amount - 1);
+                }
+            }
+        }
+
+        private static void AddRewardItemsToInventory(GameSession session, RecipeMetadata recipe)
+        {
+            // award items
+            List<RecipeItem> result = recipe.GetResult();
+            for (int i = 0; i < result.Count; i++)
+            {
+                Item rewardItem = new(result.ElementAt(i).Id)
+                {
+                    Rarity = result.ElementAt(i).Rarity, Amount = result.ElementAt(i).Amount
+                };
+                InventoryController.Add(session, rewardItem, true);
+            }
+
+            // add mastery exp
+            session.Player.Levels.GainMasteryExp(recipe.MasteryType, recipe.RewardMastery);
+        }
+
+        private static bool PlayerHasAllIngredients(GameSession session, RecipeMetadata recipe)
+        {
+            List<Item> playerInventoryItems = new(session.Player.Inventory.Items.Values);
+            List<RecipeItem> ingredients = recipe.GetIngredients();
+
+            for (int i = 0; i < ingredients.Count; i++)
+            {
+                RecipeItem ingredient = ingredients.ElementAt(i);
+                Item item = playerInventoryItems.FirstOrDefault(x => x.Id == ingredient.Id);
+                if (item != null)
+                {
+                    return item.Amount >= ingredient.Amount;
+                }
+            }
+
+            return false;
+        }
+
+        private static int[] GetDigits(int source)
```

Comment:
This also seems like a utility function not specific to this context (even if this is the only time we have needed this), perhaps this belongs in another file?